auth/src/dal.rs

```diff
@@ -0,0 +1,180 @@
+};
+use async_trait::async_trait;
+use shuttle_common::ApiKey;
+use sqlx::{
+    migrate::{MigrateDatabase, Migrator},
+    sqlite::{SqliteConnectOptions, SqliteJournalMode},
+    Row, SqlitePool,
+};
+use tracing::{error, info};
+
+pub static MIGRATIONS: Migrator = sqlx::migrate!("./migrations");
+
+#[derive(Error, Debug)]
+pub enum DalError {
+    Sqlx(#[from] sqlx::Error),
+    UserNotFound,
+}
+
+// We are not using the `thiserror`'s `#[error]` syntax to prevent sensitive details from bubbling up to the users.
+// Instead we are logging it as an error which we can inspect.
+impl fmt::Display for DalError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let msg = match self {
+            DalError::Sqlx(error) => {
+                error!(error = error.to_string(), "database request failed");
+
+                "failed to interact with auth database"
+            }
+            DalError::UserNotFound => "user not found",
+        };
+
+        write!(f, "{msg}")
+    }
+}
+
+#[async_trait]
+pub trait Dal {
+    /// Create a new account
+    async fn create_user(&self, name: AccountName, tier: AccountTier) -> Result<User, DalError>;
+
+    /// Get an account by [AccountName]
+    async fn get_user(&self, name: AccountName) -> Result<User, DalError>;
+
+    /// Get an account by [ApiKey]
+    async fn get_user_by_key(&self, key: ApiKey) -> Result<User, DalError>;
+
+    /// Reset an account's [ApiKey]
+    async fn reset_key(&self, name: AccountName) -> Result<(), DalError>;
+}
+
+pub struct Sqlite {
+    pool: SqlitePool,
+}
+
+impl Sqlite {
+    /// This function creates all necessary tables and sets up a database connection pool.
+    pub async fn new(path: &str) -> Self {
+        if !Path::new(path).exists() {
+            sqlx::Sqlite::create_database(path).await.unwrap();
+        }
+
+        info!(
+            "state db: {}",
+            std::fs::canonicalize(path).unwrap().to_string_lossy()
+        );
+
+        // We have found in the past that setting synchronous to anything other than the default (full) breaks the
+        // broadcast channel in deployer. The broken symptoms are that the ws socket connections won't get any logs
+        // from the broadcast channel and would then close. When users did deploys, this would make it seem like the
+        // deploy is done (while it is still building for most of the time) and the status of the previous deployment
+        // would be returned to the user.
+        //
+        // If you want to activate a faster synchronous mode, then also do proper testing to confirm this bug is no
+        // longer present.
+        let sqlite_options = SqliteConnectOptions::from_str(path)
+            .unwrap()
+            .journal_mode(SqliteJournalMode::Wal);
+
+        let pool = SqlitePool::connect_with(sqlite_options).await.unwrap();
+
+        Self::from_pool(pool).await
+    }
+
+    pub async fn insert_admin(&self, account_name: &str, key: Option<&str>) {
```

Comment:
To call `create_user` we need to already be an admin user, so init is a utility to do that before we start. Upgrading and downgrading a user is a good point, we don't have that functionality currently. I think we can wait though until we need it, WDYT?

Reply:
> To call create_user we need to already be an admin user, so init is a utility to do that before we start.

I don't see where we're checking that the admin exists within the method implementation. Maybe we're checking in the scope of the caller? If yes, it doesn't really matter who calls the `create_user` method (to create a regular user or an admin).

Reply:
Oh, I think I get it now. I am not suggesting going that route to create an admin user, but use Dal's `create_user` method instead of `insert_admin` (to dedup the code basically).

Reply:
Yes, that means we would need to always have to create an `admin` user by generating a random key. I don't see this necessarily as limiting. In case we'll need a custom key then we'll go your route, right?

Reply:
I am now thinking maybe we should go back to the previous design and have the ability to insert an admin be separate from the sqlite struct entirely. I moved it into the sqlite struct since it wasn't practical to have it separate while following the resource-recorder design, but that means it can be called after startup. And the resource-recorder doesn't have the insert_admin requirement. :thinking:

Reply:
I will address this as discussed in the grpc PR